### PR TITLE
[hab] Avoid install failure when running concurrently

### PR DESCRIPTION
**Pull request by stevendanna**
Commit b688d8ae25d9f38b4053fd29372e2733c9254327 changed the install
process to use a temporary directory + rename strategy. The goal of
that change was to avoid incomplete package installations. However,
that change had an unintended consequence of breaking concurrent
installs.  While concurrent installs were not previously not
guaranteed to work, they often would by chance; now, they never work
because rename(2) fails when the target is a non-empty directory.

This change checks if the package we are installing happens to exist
after an error from rename(2). If it does, we proceed as if there were
no error. If it doesn't we raise the error.

I've opted not to catch a specific error here to avoid platform
specific code. Even on linux alone the manual page for this case says
we might get ENOTEMPTY or EEXIST. I can be convinced we should
check for a specific error though.

Signed-off-by: Steven Danna <steve@chef.io>

### CODE REVIEW

File: components/common/src/command/package/install.rs [hunk 1]
```diff
@@ -732,10 +732,28 @@ impl<'a> InstallTask<'a> {
                 let temp_dir = temp_package_directory(real_install_path)?;
                 let temp_install_path = &pkg_install_path(ident, Some(temp_dir.path()));
                 artifact.unpack(Some(temp_dir.path()))?;
-                fs::rename(temp_install_path, real_install_path)?;
+
+                match fs::rename(temp_install_path, real_install_path) {
```
Comment (fnichol) [inline components/common/src/command/package/install.rs]: Not a big deal, but you could avoid the skipped `Ok(…)` with an `if let` like:

```suggestion
                if let Err(e) = fs::rename(temp_install_path, real_install_path) {
                    // ...
                }
```

**Comment by fnichol** [comment]
@stevendanna Looks like you might have to pass it through `rustfmt` from the Rust 1.31.0 era release?

File: components/common/src/command/package/install.rs (hunk 1 above)
Comment (baumanj) [reply components/common/src/command/package/install.rs]: I think you can replace this whole construct with:
```rust
                fs::rename(temp_install_path, real_install_path).or_else(|e| {
                    debug!("rename failed with {:?}, checking for installed package", e);
                    PackageInstall::load(ident.as_ref(), Some(self.fs_root_path)).map(|_| ())
                })?;
```

**Comment by baumanj** [comment]
The travis errors look like just rustfmt stuff. I've re-started the AppVeyor wheel of misfortune.

File: components/common/src/command/package/install.rs (hunk 1 above)
Comment (stevendanna) [reply components/common/src/command/package/install.rs]: How do you feel about the `if let` change I've pushed vs the or_else and map.  

To be honest, I don't think `.map(|_| ())})?;` adds much to the clarity of this code.  For example, the current code is returning the error from the rename if the package lookup fails. Does this example still do that?  To figure it out I have to pretty carefully parse rather dense code.

I'm happy to change it if y'all feel strongly about it though.

**Comment by stevendanna** [comment]
Hrm.

```
failures:
---- rumor::departure::two_members_share_departures stdout ----
thread 'rumor::departure::two_members_share_departures' panicked at 'assertion failed: u.is_some()', components/butterfly/tests/rumor/departure.rs:27:58
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::print
             at libstd/sys_common/backtrace.rs:71
             at libstd/sys_common/backtrace.rs:59
   2: std::panicking::default_hook::{{closure}}
             at libstd/panicking.rs:211
   3: std::panicking::default_hook
             at libstd/panicking.rs:221
   4: std::panicking::rust_panic_with_hook
             at libstd/panicking.rs:476
   5: std::panicking::begin_panic
             at libstd/panicking.rs:410
   6: integration::rumor::departure::two_members_share_departures::{{closure}}
             at components/butterfly/tests/rumor/departure.rs:27
   7: <habitat_butterfly::rumor::RumorStore<T>>::with_rumor
             at ./src/rumor/mod.rs:421
   8: integration::rumor::departure::two_members_share_departures
             at components/butterfly/tests/rumor/departure.rs:25
   9: integration::rumor::departure::two_members_share_departures::{{closure}}
             at components/butterfly/tests/rumor/departure.rs:20
  10: core::ops::function::FnOnce::call_once
             at libcore/ops/function.rs:238
  11: <F as alloc::boxed::FnBox<A>>::call_box
             at libtest/lib.rs:1468
             at libcore/ops/function.rs:238
             at liballoc/boxed.rs:672
  12: __rust_maybe_catch_panic
             at libpanic_unwind/lib.rs:102
failures:
    rumor::departure::two_members_share_departures
test result: FAILED. 20 passed; 1 failed; 2 ignored; 0 measured; 0 filtered out
````

**Comment by raskchanky** [comment]
@stevendanna Yeah, FWIW, I've been seeing failures from various pieces of the butterfly integration test suite in my PRs too.  From what I can tell, travis just has a hard time running these tests, for reasons I don't fully understand.  Part of me wants to disable all the butterfly integration tests in travis until we can dig in and sort it out for good.

File: components/common/src/command/package/install.rs (hunk 1 above)
Comment (baumanj) [reply components/common/src/command/package/install.rs]: I think the `if let` is an improvement, but I find it still less idiomatic than the rename/or_else/load change that I suggested. I agree that it's a bit dense, but that may be a matter of rust familiarity. Conceptually, it's a perfect match for what we're trying to say here: try a `rename`, `or else` if that fails, do a `load`. I think the tradeoff with density is that this more FP-oriented approach makes a large class of logic errors less likely.

You're right that the `.map(|_| ())` doesn't add anything; it's just required to discard the information that `load` returns in the success case. It's rather ugly, and we could avoid it if we added an appropriate `From` implementation to convert the `Result` type. How much better does this look to you?

```rust
fs::rename(temp_install_path, real_install_path).or_else(|e| {
    debug!("rename failed with {:?}, checking for installed package", e);
    PackageInstall::load(ident.as_ref(), Some(self.fs_root_path))
})?;
```

All that said, if you prefer the version you have now, I'm happy to go with it; I don't want to be imperious about style.

**Comment by stevendanna** [comment]
Wondering if someone can mash retry on the Travis and AppVeyor jobs.  I'm thinking both of those failures may be transient

**Comment by baumanj** [comment]
Retry mashed

**Pull request merged**